tiatoolbox/tools/graph.py

```diff
@@ -0,0 +1,275 @@
+            Features which have a range less than this are ignored.
+            Defaults to 1e-4.
+
+    Returns:
+        dict: A dictionary defining a graph for serialisation (e.g.
+        JSON) or converting into a torch-geometric Data object.
+            - :class:`numpy.ndarray` - x:
+                Features of each node (mean of features in a cluster).
+            - :class:`numpy.ndarray` - edge_index:
+                Edge index matrix defining connectivity.
+            - :py:obj:`Number` - y:
+                The label of the graph.
+
+    Example:
+        >>> points = np.random.rand(99, 2) * 1000
+        >>> features = np.array([
+        ...     np.random.rand(11) * n
+        ...     for n, _ in enumerate(points)
+        ... ])
+        >>> graph_dict = build_graph_dict(points, features)
+
+    """
+    # Remove features which do not change significantly between patches
+    feature_ranges = np.max(features, axis=0) - np.min(features, axis=0)
+    where_significant = feature_ranges > feature_range_thresh
+    features = features[:, where_significant]
+
+    # Build a kd-tree and find neighbours within a radius
+    kdtree = sKDTree(points)
+    neighbour_indexes, neighbour_distances = kdtree.query_radius(
+        X=points,
+        r=neighbour_search_radius,
+        return_distance=True,
+        sort_results=True,
+    )
+
+    # Initialise an empty 1-D condensed distance matrix.
+    # For information on condensed distance matrices see:
+    # noqa - https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist
+    # noqa - https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html
+    condensed_distance_matrix = np.zeros(int(len(points) * (len(points) - 1) / 2))
+
+    # Find the similarity between pairs of patches
+    index = 0
+    for i in range(len(points) - 1):
+        # Called f in the paper
+        neighbour_feature_similarities = np.exp(
+            -lambda_f
+            * np.linalg.norm(features[i] - features[neighbour_indexes[i]], axis=1)
+        )
+        # Called d in paper
+        neighbour_distance_similarities = np.exp(-lambda_d * neighbour_distances[i])
+        # 1 - product of similarities (1 - fd)
+        # (1 = most un-similar 0 = most similar)
+        neighbour_similarities = (
+            1 - neighbour_feature_similarities * neighbour_distance_similarities
+        )
+        # Initialise similarity of coordinate i vs all coordinates to 1
+        # (most un-similar).
+        i_vs_all_similarities = np.ones(len(points))
+        # Set the neighbours similarity to calculated values (similarity/fd)
+        i_vs_all_similarities[neighbour_indexes[i]] = neighbour_similarities
+        i_vs_all_similarities = i_vs_all_similarities[i + 1 :]
+        condensed_distance_matrix[
+            index : index + len(i_vs_all_similarities)
+        ] = i_vs_all_similarities
+        index = index + len(i_vs_all_similarities)
```

Comment:
How long would this take? It's better to support
- pre-generated distance matrix NxM, this is what sklearn does for Agglomerative Clustering
- directly using pdist when NxM is small
- and this last resort above

The main concern will usually the size of the matrix, but looping like this wont utilize numpy parallel very well. Additionally, this seems to be 2 ways linkage between points, so I think there are some wasted processing here.

At the moment this is ok as this is more like a prototype rather than an optimal implementation, you may want to revisit this later though.

Reply:
Also, this is not a 2-way linkage between points. No direction between two nodes. For example, we have a list of nodes [1,2,3,4,5]. Then the length of this matrix 'condensed_distance_matrix' is 10 (5x(5-1)/2).

Reply:
@wenqi006 Not the best way to express non-duplicated pairing but ok for now. I didnt realized that.
@John-P It would be best if this is under a class, as it seems to be dedicated to 1 form of spatial clustering. tbhn, likely wont be in time for the draft so just go with this for now